lib/os/onoff.c

```diff
@@ -0,0 +1,542 @@
+		       int res)
+{
+	while (!sys_slist_is_empty(list)) {
+		sys_snode_t *node = sys_slist_get_not_empty(list);
+		struct onoff_client *cli =
+			CONTAINER_OF(node,
+				     struct onoff_client,
+				     node);
+
+		notify_one(srv, cli, res);
+	}
+}
+
+static void onoff_stop_notify(struct onoff_service *srv,
+			      int res);
+
+static void onoff_start_notify(struct onoff_service *srv,
+			       int res)
+{
+	bool stop = false;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+
+	/* Can't have a queued releaser during start */
+	__ASSERT_NO_MSG(srv->releaser == NULL);
+
+	/* If the start failed log an error and leave the rest of the
+	 * state in place for diagnostics.
+	 *
+	 * If the start succeeded record a reference for all clients
+	 * and set the state to ON.  But if there are no clients,
+	 * initiate a transition to off.
+	 *
+	 * In either case reset the client queue and notify all
+	 * clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+	} else {
+		sys_snode_t *node;
+
+		SYS_SLIST_FOR_EACH_NODE(&clients, node) {
+			srv->refs += 1;
+		}
+		set_service_state(srv, SERVICE_STATE_ON);
+		stop = (srv->refs == 0);
+		if ((stop || k_is_in_isr())
+		    && ((srv->flags & ONOFF_SERVICE_STOP_WAITS) != 0U)) {
+			srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+			stop = false;
+		}
+	}
+
+	sys_slist_init(&srv->clients);
+
+	k_spin_unlock(&srv->lock, key);
+
+	notify_all(srv, &clients, res);
+	if (stop) {
+		srv->stop(srv, onoff_stop_notify);
+	}
+}
+
+int z_impl_onoff_request(struct onoff_service *srv,
+			 struct onoff_client *cli)
+{
+	bool add_client = false;        /* add client to pending list */
+	bool start = false;             /* invoke start transition */
+	bool notify = false;            /* do client notification */
+	bool no_wait = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv != 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	if (srv->refs == UINT16_MAX) {
+		rv = -EAGAIN;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_TO_OFF:
+		/* Reject if no-wait because operation will not be
+		 * completed when this call returns).
+		 */
+		if (no_wait) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Queue to start after release */
+		__ASSERT_NO_MSG(srv->releaser != NULL);
+		add_client = true;
+		rv = 3;
+		break;
+	case SERVICE_STATE_OFF:
+		/* Reject if in a no-wait context and start could
+		 * wait.
+		 */
+		if ((no_wait || k_is_in_isr())
+		    && ((srv->flags & ONOFF_SERVICE_START_WAITS) != 0U)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Start with first request while off */
+		__ASSERT_NO_MSG(srv->refs == 0);
+		set_service_state(srv, SERVICE_STATE_TO_ON);
+		start = true;
+		add_client = true;
```

Comment:
i think that this one is not right. In case `no wait` client is triggering start (first requestor) then it should not be added to the list. Should be something like `add_client = !no_wait`.

Reply:
No, this is right.  It's already been checked that the start call will complete without waiting, and part of the start is invoking `onoff_notify_start()` which notifies everything in the client list.

The only time you don't record the client is when the service is already in the desired state so the transition function isn't invoked.

Reply:
but if `no wait` client is created on stack it cannot be enqueued, it just wants to trigger start and increase ref counter

Reply:
in case of clock driver, sys timer wants to request LF clock and it does not care when it is started. cannot use signal because it's before kernel and callback would require to keep the client variable. i want to use stack-allocated, no-wait client there.

Reply:
It can be enqueued, because the start function will complete, and the list will be processed and cleared, before the `onoff_service_request()` call returns.  Completion of the notification is what transfers control of the client state back to the application at which point the structure can be cleared.

Reply:
> It can be enqueued, because the start function will complete, and the list will be processed and cleared, before the onoff_service_request() call returns.

No, that's not true. If it is the first request then start function will trigger clock starting which will complete in hundreds of milliseconds. Since it was in `SERVICE_STATE_OFF` it has `add_client=true` and `start=true`. Client is put into the list and start function is called and function leaves. ref is not incremented. In start notify (called from clock isr) list is processed but client on that list is no longer valid because it was allocated on the stack.

Reply:
There's some confusion about no-wait.  No-wait means "do not execute this unless the entire operation will be complete at the time the function returns."  You cannot successfully request a no-wait start when the transition will wait.